modules/data.atmosphere/R/download.NOAA_GEFS_downscale.R

```diff
@@ -0,0 +1,341 @@
+##' data frame contains information about one file.
+##'
+##' @param outfolder Directory where results should be written
+##' @param start_date, end_date Range of dates/times to be downloaded (default assumed time of day is 0:00, midnight)
+##' @param lat site latitude in decimal degrees
+##' @param lon site longitude in decimal degrees
+##' @param sitename The unique ID given to each site. This is used as part of the file name.
+##' @param overwrite logical. Download a fresh version even if a local file with the same name already exists?
+##' @param verbose logical.  Print additional debug information.  Passed on to functions in the netcdf4 package to provide debugging info.
+##' @param ... Other arguments, currently ignored
+##' @export
+##' 
+##' @examples 
+##' \dontrun{
+##'  download.NOAA_GEFS(outfolder="~/Working/results", lat.in= 45.805925, lon.in = -90.07961, sitename="US-WCr")
+##' }
+##' 
+##' @author Katie Zarada - modified code from Luke Dramko and Laura Puckett
+##' 
+##' 
+
+
+download.NOAA_GEFS_downscale <- function(outfolder, lat.in, lon.in, sitename, start_date = Sys.time(), end_date = (as.POSIXct(start_date, tz="UTC") + lubridate::days(16)),
+                                         overwrite = FALSE, verbose = FALSE, ...) {
+  
+  start_date <- as.POSIXct(start_date, tz = "UTC")
+  end_date <- as.POSIXct(end_date, tz = "UTC")
+  
+  #It takes about 2 hours for NOAA GEFS weather data to be posted.  Therefore, if a request is made within that 2 hour window,
+  #we instead want to adjust the start time to the previous forecast, which is the most recent one avaliable.  (For example, if
+  #there's a request at 7:00 a.m., the data isn't up yet, so the function grabs the data at midnight instead.)
+  if (abs(as.numeric(Sys.time() - start_date, units="hours")) <= 2) {
+    start_date = start_date - lubridate::hours(2)
+    end_date = end_date - lubridate::hours(2)
+  }
+  
+  #Date/time error checking - Checks to see if the start date is before the end date
+  if (start_date > end_date) {
+    PEcAn.logger::logger.severe("Invalid dates: end date occurs before start date")
+  } else if (as.numeric(end_date - start_date, units="hours") < 6) { #Done separately to produce a more helpful error message.
+    PEcAn.logger::logger.severe("Times not far enough appart for a forecast to fall between them.  Forecasts occur every six hours; make sure start 
+                                and end dates are at least 6 hours appart.")
+  }
+  
+  #Set the end forecast date (default is the full 16 days)
+  if (end_date > start_date + lubridate::days(16)) {
+    end_date = start_date + lubridate::days(16)
+  }
+  
+  #Round the starting date/time down to the previous block of 6 hours.  Adjust the time frame to match.
+  forecast_hour = (lubridate::hour(start_date) %/% 6) * 6 #Integer division by 6 followed by re-multiplication acts like a "floor function" for multiples of 6
```

Comment:
Just FYI, see also the `lubridate::floor_date` function. If I'm reading this correctly, this should be analogous to:

```r
lubridate::floor_date(start_date, "6 hours")
```

Reply:
This is from Luke's original code and he pulls out just the hour here and does other things with the hour below. The lubridate::floor returns the full date